### ISSUE TITLE
Bug for distributed wrapper regarding cross batch memory loss 

### ISSUE THREAD

**Issue by zhaoyuac09** [KevinMusgrave/pytorch-metric-learning#639]
First of all, I really appreciated this repo. Thank you very much for the contribution!  However, there are 2 functions will not work logically, in distributed.py for the loss and miner wrappers: gather_emb_and_ref and gather_enqueue_mask.  
Let's take gather_enqueue_mask for example: 
```
def gather_enqueue_mask(enqueue_mask, device):
    if enqueue_mask is None:
        return enqueue_mask
    enqueue_mask = c_f.to_device(enqueue_mask, device=device)
    return torch.cat([enqueue_mask, all_gather(enqueue_mask)], dim=0)

def all_gather(x):
    world_size = torch.distributed.get_world_size()
    if world_size > 1:
        rank = torch.distributed.get_rank()
        x_list = [torch.ones_like(x) for _ in range(world_size)]
        torch.distributed.all_gather(x_list, x.contiguous())
        # remove curr rank
        x_list.pop(rank)
        return torch.cat(x_list, dim=0)
    return None
```

the all_gather function poped the rank, which will be different int on different GPUs, then torch cat the current enqueue_mask. Then the order Of the all gathered masks will not be guaranteed the same.  When using cross batch memory losses, the embedding_memory will end up different on different GPUs, which I have already confirmed running some testing function.

**Comment by zhaoyuac09** [comment]
Thank you @KevinMusgrave. I would be happy to create a pull request later after I finish more testing cases here.  I also found a potential issue which are mentioned here: https://github.com/KevinMusgrave/pytorch-metric-learning/issues/562#issuecomment-1589636568 for the loss scaling by world_size. If later I have succeeded all testing cases, I will wrap up all changes and open a pull request.   
Another issue is, when cross batch memory loss is wrapped with the distributed wrapper, miner cannot be wrapped again since miner will already have access of all embs, labels, etc after all gathering in the loss wrapper (https://github.com/KevinMusgrave/pytorch-metric-learning/blob/c38c07c0587bad7c463ae98293d2978e931f0ae6/src/pytorch_metric_learning/utils/distributed.py#L155). 

I believe your repo is really nice and almost there for distributed training support. Thanks for the nice repo and let's make it even better.

**Comment by lolongcovas** [comment]
I am facing the same issue. @KevinMusgrave have you reviewed @zhaoyuac09 PR?

**Comment by KevinMusgrave** [comment]
Here's the test file: https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/tests/utils/test_distributed.py
